builtin.c

```diff
@@ -0,0 +1,105 @@
+/*    builtin.c
+ *
+ *    Copyright (C) 2021 by Paul Evans and others
+ *
+ *    You may distribute under the terms of either the GNU General Public
+ *    License or the Artistic License, as specified in the README file.
+ *
+ */
+
+/* This file contains the code that implements functions in perl's "builtin::"
+ * namespace
+ */
+
+#include "EXTERN.h"
+#include "perl.h"
+
+#include "XSUB.h"
+
+XS(XS_builtin_true);
+XS(XS_builtin_true)
+{
+    dXSARGS;
+    if(items)
+        croak_xs_usage(cv, "");
+    XSRETURN_YES;
+}
+
+XS(XS_builtin_false);
+XS(XS_builtin_false)
+{
+    dXSARGS;
+    if(items)
+        croak_xs_usage(cv, "");
+    XSRETURN_NO;
+}
+
+XS(XS_builtin_isbool);
+XS(XS_builtin_isbool)
+{
+    dXSARGS;
+    if(items != 1)
+        croak_xs_usage(cv, "sv");
+
+    SV *sv = ST(0);
+    if(SvIsBOOL(sv))
+        XSRETURN_YES;
+    else
+        XSRETURN_NO;
+}
+
+XS(XS_builtin_import);
+XS(XS_builtin_import)
+{
+    dXSARGS;
+    int i;
+
+    if(!PL_compcv)
+        Perl_croak(aTHX_
+                "builtin::import can only be called at compiletime");
+
+    /* We need to have PL_comppad / PL_curpad set correctly for lexical importing */
+    ENTER;
+    SAVESPTR(PL_comppad_name); PL_comppad_name = PadlistNAMES(CvPADLIST(PL_compcv));
+    SAVESPTR(PL_comppad);      PL_comppad      = PadlistARRAY(CvPADLIST(PL_compcv))[1];
+    SAVESPTR(PL_curpad);       PL_curpad       = PadARRAY(PL_comppad);
+
+    for(i = 1; i < items; i++) {
+        SV *sym = ST(i);
+        if(strEQ(SvPV_nolen(sym), "import")) goto unavailable;
+
+        /* TODO: I've no idea why the newSVpvf macro isn't visible here */
```

Comment:
Odd, this compiles fine for me with the short versions

Reply:
Weird. When I try it I get:

```
In file included from perl.h:5744,
                 from builtin.c:15:
builtin.c: In function ‘XS_builtin_import’:
builtin.c:71:34: warning: implicit declaration of function ‘newSVpvf’; did you mean ‘newSVpv’? [-Wimplicit-function-declaration]
   71 |         SV *ampname = sv_2mortal(newSVpvf("&%" SVf, SVfARG(sym)));
      |                                  ^~~~~~~~
embed.h:578:46: note: in definition of macro ‘sv_2mortal’
  578 | #define sv_2mortal(a)  Perl_sv_2mortal(aTHX_ a)
      |                                              ^
builtin.c:71:34: warning: passing argument 2 of ‘Perl_sv_2mortal’ makes pointer from integer without a cast [-Wint-conversion]
   71 |         SV *ampname = sv_2mortal(newSVpvf("&%" SVf, SVfARG(sym)));
      |                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                  |
      |                                  int
...
cc -fstack-protector-strong -L/usr/local/lib -o miniperl \
    opmini.o perlmini.o universalmini.o  gv.o toke.o perly.o pad.o regcomp.o dump.o util.o mg.o reentr.o mro_core.o keywords.o builtin.o hv.o av.o run.o pp_hot.o sv.o pp.o scope.o pp_ctl.o pp_sys.o doop.o doio.o regexec.o utf8.o taint.o deb.o globals.o perlio.o numeric.o mathoms.o locale.o pp_pack.o pp_sort.o caretx.o dquote.o time64.o  miniperlmain.o  -lpthread -lnsl -ldl -lm -lcrypt -lutil -lc 
/usr/bin/ld: builtin.o: in function `XS_builtin_import':
/home/leo/src/bleadperl/perl/builtin.c:71: undefined reference to `newSVpvf'
/usr/bin/ld: /home/leo/src/bleadperl/perl/builtin.c:72: undefined reference to `newSVpvf'
collect2: error: ld returned 1 exit status
```

Reply:
It seems that `embed.h` defines `newSVpvf` only under non-MULTIPLICITY builds:

```c
 394 #ifndef MULTIPLICITY
 395 #define newSVpvf                Perl_newSVpvf
 396 #endif
 397 #define newSVpvn(a,b)           Perl_newSVpvn(aTHX_ a,b)
 398 #define newSVpvn_flags(a,b,c)   Perl_newSVpvn_flags(aTHX_ a,b,c)
 399 #define newSVpvn_share(a,b,c)   Perl_newSVpvn_share(aTHX_ a,b,c)
```